contrib_bots/lib/giveaway.py

```diff
@@ -0,0 +1,215 @@
+    command to work
+    cancel - cancel a giveaway (can only be run by the person who first
+    initialised the giveaway
+
+    This plugin was created by Daniel O'Brien
+    as part of Google Code-In 2016-2017
+    '''
+    def __init__(self):
+        self.giveaway_array = []
+        self.giveaway_on = False
+        self.giveaway_name = ""
+        self.giveaway_time = 0
+        self.giveaway_creator = ""
+        self.giveaway_start_time = 0
+        self.giveaway_group = ""
+        self.giveaway_subject = ""
+        self.giveaway_client = ""
+        self.giveaway_timeMins = 0
+        self.giveaway_stage = 0
+        self.one_minute_warning = False
+        self.giveaway_winner = ""
+        self.giveaway_winner_contact_address = []
+        self.winner_index = 0
+        self.all_warnings_fired = False
+
+    def time_update(self, new_content):
+        if self.giveaway_on:
+            self.giveaway_client.send_message(dict(
+                type='stream',
+                to=self.giveawayStream,
+                subject=self.giveaway_subject,
+                content=new_content,
+            ))
+            threading.Timer(6.0, self.check_time).start()
+
+    def time_warning(self):
+        if self.giveaway_stage == 0:
+            self.numerator = 1
+            self.denominator = 4
+        elif self.giveaway_stage == 1:
+            self.numerator = 1
+            self.denominator = 2
+        elif self.giveaway_stage == 2:
+            self.numerator = 3
+            self.denominator = 4
+        elif self.giveaway_stage == 0:
+            self.all_warnings_fired = True
+        if time.time() > ((self.giveaway_start_time + self.giveaway_time)-60):
+            if self.giveaway_timeMins > 1.9 and not self.one_minute_warning:
+                new_content = 'Giveaway ends in 1 minute'
+                self.time_update(new_content)
+                self.one_minute_warning = True
+        elif time.time() > (self.giveaway_start_time + (old_div(self.giveaway_time, self.denominator)*self.numerator)):
+            if not self.all_warnings_fired:
+                new_content = ('{}/{} of giveaway time elapsed, {} minutes remaining'
+                               .format(self.numerator, self.denominator, str((old_div(self.giveaway_timeMins, 4))*3)))
+                self.time_update(new_content)
+                self.giveaway_stage = self.giveaway_stage + 1
+
+    def check_time(self):
+        if time.time() > (self.giveaway_start_time + self.giveaway_time):
+            print("Giveaway ended")
+            try:
+                self.giveaway_winner = random.choice(self.giveaway_array)
+                self.winner_index = self.giveaway_array.index(self.giveaway_winner)
+                new_content = 'The giveaway has now ended... The winner is... {}!'.format(self.giveaway_winner)
+                direct_content = """Congratulations!, you recently partook in the giveaway called {} and won!
```

Comment:
I'll double check but last time I tried, zulip took the string literally and just word wrapped it around several lines